# Day 4: Giant Squid

## Part one

You're already almost 1.5km (almost a mile) below the surface of the ocean, already so deep that you can't see any sunlight. What you can see, however, is a giant squid that has attached itself to the outside of your submarine.

Maybe it wants to play bingo?

Bingo is played on a set of boards each consisting of a 5x5 grid of numbers. Numbers are chosen at random, and the chosen number is marked on all boards on which it appears. (Numbers may not appear on all boards.) If all numbers in any row or any column of a board are marked, that board wins. (Diagonals don't count.)

The submarine has a bingo subsystem to help passengers (currently, you and the giant squid) pass the time. It automatically generates a random order in which to draw numbers and a random set of boards (your puzzle input). For example:

`7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1`

```
22 13 17 11  0
 8  2 23  4 24
21  9 14 16  7
 6 10  3 18  5
 1 12 20 15 19

 3 15  0  2 22
 9 18 13 17  5
19  8  7 25 23
20 11 10 24  4
14 21 16 12  6

14 21 17 24  4
10 16 15  9 19
18  8 23 26 20
22 11 13  6  5
 2  0 12  3  7
```

After the first five numbers are drawn (`7`, `4`, `9`, `5`, and `11`), there are no winners, but the boards are marked as follows (shown here adjacent to each other to save space):

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

After the next six numbers are drawn (`17`, `23`, `2`, `0`, `14`, and `21`), there are still no winners:

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

Finally, `24` is drawn:

```
22 13 17 11  0         3 15  0  2 22        14 21 17 24  4
 8  2 23  4 24         9 18 13 17  5        10 16 15  9 19
21  9 14 16  7        19  8  7 25 23        18  8 23 26 20
 6 10  3 18  5        20 11 10 24  4        22 11 13  6  5
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  7
```

At this point, the third board **wins** because it has at least one complete row or column of marked numbers (in this case, the entire top row is marked: `14 21 17 24 4`).

The **score** of the winning board can now be calculated. Start by finding the **sum of all unmarked numbers** on that board; in this case, the sum is `188`. Then, multiply that sum by **the number that was just called** when the board won, `24`, to get the final score, `188 * 24 = 4512`.

To guarantee victory against the giant squid, figure out which board will win first. **What will your final score be if you choose that board?**

In [1]:
from collections import namedtuple

In [2]:
def load_input(filepath: str='./data/input.txt') -> tuple:
    with open(filepath, 'r') as f:
        draws = tuple(int(el) for el in next(f).rstrip().split(','))
        _ = next(f)
        lines = [l.rstrip() for l in f]

    def build_board(lines):
        return tuple(
            tuple(
                {'value': int(item), 'drawn': False} for item in line.split(' ') if item) for line in lines
            )

    boards = tuple(build_board(lines[i:i + 5]) for i in range(0, len(lines), 6))

    return draws, boards

draws, boards = load_input()
print(draws, boards[0])

(37, 60, 87, 13, 34, 72, 45, 49, 61, 27, 97, 88, 50, 30, 76, 40, 63, 9, 38, 67, 82, 6, 59, 90, 99, 54, 11, 66, 98, 23, 64, 14, 18, 4, 10, 89, 46, 32, 19, 5, 1, 53, 25, 96, 2, 12, 86, 58, 41, 68, 95, 8, 7, 3, 85, 70, 35, 55, 77, 44, 36, 51, 15, 52, 56, 57, 91, 16, 71, 92, 84, 17, 33, 29, 47, 75, 80, 39, 83, 74, 73, 65, 78, 69, 21, 42, 31, 93, 22, 62, 24, 48, 81, 0, 26, 43, 20, 28, 94, 79) (({'value': 66, 'drawn': False}, {'value': 78, 'drawn': False}, {'value': 7, 'drawn': False}, {'value': 45, 'drawn': False}, {'value': 92, 'drawn': False}), ({'value': 39, 'drawn': False}, {'value': 38, 'drawn': False}, {'value': 62, 'drawn': False}, {'value': 81, 'drawn': False}, {'value': 77, 'drawn': False}), ({'value': 9, 'drawn': False}, {'value': 73, 'drawn': False}, {'value': 25, 'drawn': False}, {'value': 97, 'drawn': False}, {'value': 99, 'drawn': False}), ({'value': 87, 'drawn': False}, {'value': 80, 'drawn': False}, {'value': 19, 'drawn': False}, {'value': 1, 'drawn': False}, {'value': 71, '

In [3]:
DrawnElem = namedtuple('DrawnElem', 'el board_num line_num col_num')

def bingo_draw(selected_number: int, boards: tuple) -> tuple:
    # Select every elem that have been drawn
    # along with their board, line and col number
    drawn_elems = tuple(DrawnElem(boards[board_num][line_num][col_num], board_num, line_num, col_num) 
                for board_num in range(len(boards)) 
                for line_num in range(len(boards[board_num])) 
                for col_num in range(len(boards[board_num][line_num])) 
                if boards[board_num][line_num][col_num]['value'] == selected_number)

    # Set their drawn parameter to true
    for d_el in drawn_elems:
        d_el.el['drawn'] = True
    
    return drawn_elems

def bingo_win(drawn_elems: tuple, boards: tuple) -> tuple:
    for el in drawn_elems:
        # Check items of the line
        l_drawn = [int(line_elem['drawn']) 
                for line_elem in boards[el.board_num][el.line_num]]
        if sum(l_drawn) == 5:
            return True, el

        # Check items of the column
        c_drawn = [int(boards[el.board_num][line_num][el.col_num]['drawn']) 
                for line_num in range(len(boards[el.board_num][el.line_num]))]
        if sum(c_drawn) == 5:
            return True, el

    return False, None

In [4]:
draws, boards = load_input()

for sel_num in draws:
    drawn_elems = bingo_draw(sel_num, boards)

    has_win, el = bingo_win(drawn_elems, boards)
    if has_win:
        break

el

DrawnElem(el={'value': 54, 'drawn': True}, board_num=14, line_num=2, col_num=0)

In [5]:
def sum_unmarked(board):
    return sum([el['value'] for line in board for el in line if not el['drawn']])

In [6]:
sum_unmarked(boards[el.board_num]) * el.el['value']

34506

Your puzzle answer was `34506`.

## Part two

On the other hand, it might be wise to try a different strategy: let the giant squid win.

You aren't sure how many bingo boards a giant squid could play at once, so rather than waste time counting its arms, the safe thing to do is to **figure out which board will win last** and choose that one. That way, no matter which boards it picks, it will win for sure.

In the above example, the second board is the last to win, which happens after 13 is eventually called and its middle column is completely marked. If you were to keep playing until this point, the second board would have a sum of unmarked numbers equal to 148 for a final score of `148 * 13 = 1924`.

Figure out which board will win last. **Once it wins, what would its final score be?**

In [7]:
DrawnElem = namedtuple('DrawnElem', 'el board_num line_num col_num')

def bingo_draw(selected_number: int, boards: tuple, remaining_board_nums: list) -> tuple:
    # Select every elem that have been drawn
    # along with their board, line and col number
    drawn_elems = tuple(DrawnElem(boards[board_num][line_num][col_num], board_num, line_num, col_num) 
                for board_num in remaining_board_nums 
                for line_num in range(len(boards[board_num])) 
                for col_num in range(len(boards[board_num][line_num])) 
                if boards[board_num][line_num][col_num]['value'] == selected_number)

    # Set their drawn parameter to true
    for d_el in drawn_elems:
        d_el.el['drawn'] = True
    
    return drawn_elems

def bingo_win(drawn_elems: tuple, boards: tuple) -> tuple:
    ending_els = []
    for el in drawn_elems:
        # Check items of the line
        l_drawn = [int(line_elem['drawn']) 
                for line_elem in boards[el.board_num][el.line_num]]
        if sum(l_drawn) == 5:
            ending_els.append(el)
            continue

        # Check items of the column
        c_drawn = [int(boards[el.board_num][line_num][el.col_num]['drawn']) 
                for line_num in range(len(boards[el.board_num][el.line_num]))]
        if sum(c_drawn) == 5:
            ending_els.append(el)
            continue

    return (True, ending_els) if ending_els else (False, None)

In [8]:
draws, boards = load_input()
remaining_board_nums = list(range(len(boards)))

for sel_num in draws:
    drawn_elems = bingo_draw(sel_num, boards, remaining_board_nums)

    has_win, ending_els = bingo_win(drawn_elems, boards)
    if has_win:
        ended_b_num = [b_num.board_num for b_num in ending_els]
        remaining_board_nums = [el for el in remaining_board_nums if el not in ended_b_num]

    if not remaining_board_nums:
        break

el = ending_els[0]

In [9]:
sum_unmarked(boards[el.board_num]) * el.el['value']

7686

Your puzzle answer was 7686.